### PR TITLE
TypeLowering: crash when the phiNode's user is invariant_start intrinsic

### PR DESCRIPTION
**Pull request by ruiminzhao**
When the IR like this, it will crash:

  %2653 = load <4 x i32>, ptr addrspace(4) %2652, align 16
  ...
  %2654 = call ptr addrspace(7) @lgc.buffer.desc.to.ptr(<4 x i32> %2653)
  %2853 = call ptr addrspace(7) @lgc.buffer.desc.to.ptr(<4 x i32> %2852)
  %2932 = call ptr addrspace(7) @lgc.buffer.desc.to.ptr(<4 x i32> %2931)
  ...
  %.lcssa8772 = phi ptr addrspace(7) [ %2654, %branch27_true ], [ %2853, %branch27_true.1 ], [ %2932, %branch27_true.2 ]
  ...
  %3186 = call ptr @llvm.invariant.start.p7(i64 -1, ptr addrspace(7) %.lcssa8772)

Root Cause:
  1. removeUsersForInvariantStarts in BufferOpLowering hasn't considered on the phiNode.
  2. visitPhi in TypeLowering hasn't considered on the @llvm.invariant.start.p7
With 1 && A2, the "call ptr @llvm.invariant.start.p7(i64 -1, ptr addrspace(7) %.lcssa8772)" will never be added in m_instructionsToErase. 
So when erase the %.lcssa8772, it will crash when the user is not destroyed.

Solution:
  It should add the @llvm.invariant.start.p7 in visitPhi in
  TypeLowering as deleting instruction is in reverse order.
  (I have tried to add the phiNode in BufferOpLowering::removeUsersForInvariantStarts, but it will fail as llvm.invariant.start.p7 is in front of  PhiNode: %.lcssa8772 in m_instructionsToErase)

### CODE REVIEW

**Comment by nhaehnle** [comment]
Nice find. Can you please add a test case in lgc/test/Transforms/PatchBufferOp?

I have some concerns about this approach. It seems too "special" in a way. For example, the generic TypeLowering infrastructure shouldn't have to know about invariant.start and how it interacts with addrspace(7) pointers. And what if invariant.start occurs elsewhere, e.g. if it is applied to the result of a `select`?

Can we instead fix the problem by handling invariant.start differently in `BufferOpLowering`? Here's a rough idea:

* Remove `removeUsersForInvariantStarts`
* Instead, add a `visitInvariantStart` method that is registered to the visitor via the `addIntrinsic` method
* This method checks whether the operand is a buffer pointer, and if so, looks up the `DescriptorInfo`, sets `invariant` to true, and erases the invariant start

Related observation: The case of `bitcast` in `BufferOpLowering::removeUsersForInvariantStarts` should no longer be relevant due to opaque pointers.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 464052998e132baaef8bf16fde39bfd8a168740b
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by ruiminzhao** [comment]
> Nice find. Can you please add a test case in lgc/test/Transforms/PatchBufferOp?
> 
> I have some concerns about this approach. It seems too "special" in a way. For example, the generic TypeLowering infrastructure shouldn't have to know about invariant.start and how it interacts with addrspace(7) pointers. And what if invariant.start occurs elsewhere, e.g. if it is applied to the result of a `select`?
> 
> Can we instead fix the problem by handling invariant.start differently in `BufferOpLowering`? Here's a rough idea:
> 
> * Remove `removeUsersForInvariantStarts`
> * Instead, add a `visitInvariantStart` method that is registered to the visitor via the `addIntrinsic` method
> * This method checks whether the operand is a buffer pointer, and if so, looks up the `DescriptorInfo`, sets `invariant` to true, and erases the invariant start
> 
> Related observation: The case of `bitcast` in `BufferOpLowering::removeUsersForInvariantStarts` should no longer be relevant due to opaque pointers.

Thanks, Nicolai. I would like following your suggestion as current fix may only handle for the "Phi" case. It should visitInvariantStart to handle all cases.  Still a small question on the usage for addIntrinsic needs your clarify:

1.  * Instead, add a `visitInvariantStart` method that is registered to the visitor via the `addIntrinsic` method
===> Does it mean the code should like this:
  ```static const auto visitor = llvm_dialects::VisitorBuilder<PatchBufferOpImpl>()
                                  .nest(&BufferOpLowering::registerVisitors)
                                  .nest(&TypeLowering::registerVisitors)
                                  .addIntrinsic(<unsigned id>,  &BufferOpLowering::visitInvariantStart)
                                  .build(); ```

Which value should be the parameter  **<unsigned id>**  in addIntrinsic set here?

**Comment by nhaehnle** [comment]
It should be
```
 builder.addIntrinsic(Intrinsic::invariant_start, &BufferOpLowering::visitInvariantStart);
```
inside of `BufferOpLowering::registerVisitors`.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 464052998e132baaef8bf16fde39bfd8a168740b
<details><summary>CTS tests (Failed: 0/142046)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by ruiminzhao** [comment]
> It should be
> 
> ```
>  builder.addIntrinsic(Intrinsic::invariant_start, &BufferOpLowering::visitInvariantStart);
> ```
> 
> inside of `BufferOpLowering::registerVisitors`.

Following the suggestion with my latest commit,  still meet another crash:
```
0x00007FF7B9E2D38D, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x73FD38D byte(s), memcpy() + 0x10D byte(s), d:\a01\_work\43\s\src\vctools\crt\vcruntime\src\string\amd64\memcpy.asm, line 199 + 0x0 byte(s)
0x00007FF7B2E6C571, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x43C571 byte(s), llvm::detail::PunnedPointer<llvm::ilist_node_base<1> *>::asInt() + 0x41 byte(s), C:\DK\llvm-project\llvm\include\llvm\ADT\PointerIntPair.h, line 42 + 0x0 byte(s)
0x00007FF7B2E5FAC4, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x42FAC4 byte(s), llvm::detail::PunnedPointer<llvm::ilist_node_base<1> *>::operator __int64() + 0x14 byte(s), C:\DK\llvm-project\llvm\include\llvm\ADT\PointerIntPair.h, line 45 + 0x14 byte(s)
0x00007FF7B2E8B827, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x45B827 byte(s), llvm::PointerIntPair<llvm::ilist_node_base<1> *,1,unsigned int,llvm::PointerLikeTypeTraits<llvm::ilist_node_base<1> *>,llvm::PointerIntPairInfo<llvm::ilist_node_base<1> *,1,llvm::PointerLikeTypeTraits<llvm::ilist_node_base<1> *> > >::getInt() + 0x17 byte(s), C:\DK\llvm-project\llvm\include\llvm\ADT\PointerIntPair.h, line 96 + 0x17 byte(s)
0x00007FF7B2E8BFA7, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x45BFA7 byte(s), llvm::ilist_node_base<1>::isSentinel() + 0x17 byte(s), C:\DK\llvm-project\llvm\include\llvm\ADT\ilist_node_base.h, line 45 + 0x17 byte(s)
0x00007FF7B2E8BE04, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x45BE04 byte(s), llvm::ilist_node_base<1>::isKnownSentinel() + 0x14 byte(s), C:\DK\llvm-project\llvm\include\llvm\ADT\ilist_node_base.h, line 46 + 0x14 byte(s)
0x00007FF7B2FAFBE7, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x57FBE7 byte(s), llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction,1,0,void>,0,0>::operator*() + 0x17 byte(s), C:\DK\llvm-project\llvm\include\llvm\ADT\ilist_iterator.h, line 138 + 0xD byte(s)
0x00007FF7B868D371, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x5C5D371 byte(s), llvm_dialects::detail::VisitorBase::visit() + 0x261 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\imported\llvm-dialects\lib\Dialect\Visitor.cpp, line 311 + 0x72 byte(s)
0x00007FF7B331EBC8, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x8EEBC8 byte(s), llvm_dialects::Visitor<`anonymous namespace'::PatchBufferOpImpl>::visit() + 0x28 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\imported\llvm-dialects\include\llvm-dialects\Dialect\Visitor.h, line 311 + 0x0 byte(s)
0x00007FF7B331EA26, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x8EEA26 byte(s), `anonymous namespace'::PatchBufferOpImpl::run() + 0x146 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\lgc\patch\PatchBufferOp.cpp, line 117 + 0x0 byte(s)
0x00007FF7B331E7E8, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x8EE7E8 byte(s), lgc::PatchBufferOp::run() + 0xD8 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\lgc\patch\PatchBufferOp.cpp, line 83 + 0x12 byte(s)
0x00007FF7B2F202B0, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x4F02B0 byte(s), llvm::detail::PassModel<llvm::Function,lgc::PatchBufferOp,llvm::PreservedAnalyses,llvm::AnalysisManager<llvm::Function> >::run() + 0x40 byte(s), C:\DK\llvm-project\llvm\include\llvm\IR\PassManagerInternal.h, line 89 + 0x27 byte(s)
0x00007FF7B316C32B, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x73C32B byte(s), llvm::PassManager<llvm::Function,llvm::AnalysisManager<llvm::Function> >::run() + 0x16B byte(s), C:\DK\llvm-project\llvm\include\llvm\IR\PassManager.h, line 521 + 0x0 byte(s)
0x00007FF7B2F08570, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x4D8570 byte(s), llvm::detail::PassModel<llvm::Function,llvm::PassManager<llvm::Function,llvm::AnalysisManager<llvm::Function> >,llvm::PreservedAnalyses,llvm::AnalysisManager<llvm::Function> >::run() + 0x40 byte(s), C:\DK\llvm-project\llvm\include\llvm\IR\PassManagerInternal.h, line 89 + 0x27 byte(s)
0x00007FF7B97F1981, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x6DC1981 byte(s), llvm::ModuleToFunctionPassAdaptor::run() + 0x1A1 byte(s), C:\DK\llvm-project\llvm\lib\IR\PassManager.cpp, line 129 + 0x0 byte(s)
0x00007FF7B2F9D4F8, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x56D4F8 byte(s), llvm::detail::PassModel<llvm::Module,llvm::ModuleToFunctionPassAdaptor,llvm::PreservedAnalyses,llvm::AnalysisManager<llvm::Module> >::run() + 0x48 byte(s), C:\DK\llvm-project\llvm\include\llvm\IR\PassManagerInternal.h, line 89 + 0x27 byte(s)
0x00007FF7B32018EB, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x7D18EB byte(s), llvm::PassManager<llvm::Module,llvm::AnalysisManager<llvm::Module> >::run() + 0x16B byte(s), C:\DK\llvm-project\llvm\include\llvm\IR\PassManager.h, line 521 + 0x0 byte(s)
0x00007FF7B31D47D2, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x7A47D2 byte(s), `anonymous namespace'::PassManagerImpl::run() + 0x272 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\lgc\util\PassManager.cpp, line 277 + 0x42 byte(s)
0x00007FF7B323D3AD, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x80D3AD byte(s), lgc::PipelineState::generate() + 0x49D byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\lgc\state\Compiler.cpp, line 222 + 0x0 byte(s)
0x00007FF7B323CEDC, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x80CEDC byte(s), lgc::PipelineState::generate() + 0xDC byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\lgc\state\Compiler.cpp, line 154 + 0xBF byte(s)
0x00007FF7B2E2D9D1, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x3FD9D1 byte(s), Llpc::Compiler::buildPipelineInternal() + 0x17E1 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\context\llpcCompiler.cpp, line 1310 + 0x0 byte(s)
0x00007FF7B2E2BACC, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x3FBACC byte(s), Llpc::Compiler::buildComputePipelineInternal() + 0x4CC byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\context\llpcCompiler.cpp, line 1860 + 0xEB byte(s)
0x00007FF7B2E281CC, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x3F81CC byte(s), Llpc::Compiler::BuildComputePipeline() + 0x61C byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\context\llpcCompiler.cpp, line 1913 + 0x41 byte(s)
0x00007FF7B2E0B165, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x3DB165 byte(s), Llpc::StandaloneCompiler::ComputePipelineBuilder::buildComputePipeline() + 0x4B5 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\tool\llpcComputePipelineBuilder.cpp, line 136 + 0x6D byte(s)
0x00007FF7B2E0AB0A, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x3DAB0A byte(s), Llpc::StandaloneCompiler::ComputePipelineBuilder::build() + 0x5A byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\tool\llpcComputePipelineBuilder.cpp, line 63 + 0x12 byte(s)
0x00007FF7B2DBCFA0, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x38CFA0 byte(s), processInputs() + 0x7C0 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\tool\amdllpc.cpp, line 622 + 0x43 byte(s)
0x00007FF7B2DBD9A3, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x38D9A3 byte(s), <lambda_81d857a2c93921f1c1b2eea39ead1b12>::operator()() + 0x33 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\tool\amdllpc.cpp, line 724 + 0x33 byte(s)
0x00007FF7B2DD1613, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x3A1613 byte(s), Llpc::parallelFor<llvm::SmallVector<llvm::SmallVector<Llpc::StandaloneCompiler::InputSpec,2>,0> &,<lambda_81d857a2c93921f1c1b2eea39ead1b12> >() + 0x173 byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\util\llpcThreading.h, line 90 + 0x25 byte(s)
0x00007FF7B2DBD6CB, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x38D6CB byte(s), main() + 0x4AB byte(s), C:\DK\drivers\drivers\xgl\icd\api\compiler\llpc\tool\amdllpc.cpp, line 723 + 0x86 byte(s)
0x00007FF7B9E29259, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x73F9259 byte(s), invoke_main() + 0x39 byte(s), d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 79 + 0x0 byte(s)
0x00007FF7B9E2913E, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x73F913E byte(s), __scrt_common_main_seh() + 0x12E byte(s), d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 288 + 0x5 byte(s)
0x00007FF7B9E28FFE, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x73F8FFE byte(s), __scrt_common_main() + 0xE byte(s), d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl, line 331 + 0x0 byte(s)
0x00007FF7B9E292EE, C:\DK\drivers\drivers\xgl\build\compiler\llpc\Debug\amdllpc.exe(0x00007FF7B2A30000) + 0x73F92EE byte(s), mainCRTStartup() + 0xE byte(s), d:\a01\_work\43\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp, line 17 + 0x0 byte(s)
0x00007FFAD6097614, C:\Windows\System32\KERNEL32.DLL(0x00007FFAD6080000) + 0x17614 byte(s), BaseThreadInitThunk() + 0x14 byte(s)
0x00007FFAD6E826A1, C:\Windows\SYSTEM32\ntdll.dll(0x00007FFAD6E30000) + 0x526A1 byte(s), RtlUserThreadStart() + 0x21 byte(s)
```

Seems the erasing on Intrinsic::invariant_start will cause the IR change? It's still under investigation now.

**Comment by amdvlk-admin** [comment]
b9fb4e8e1de29fc613abbe483d9c40c3517d56e3 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/IR/PassManager.h:48:10: fatal error: llvm/IR/PassManagerInternal.h: No such file or directory
compilation terminated.
[1610/3067] Building CXX object compiler/llpc/llvm/lib/Transforms/Scalar/CMakeFiles/LLVMScalarOpts.dir/LoopUnrollPass.cpp.o
FAILED: compiler/llpc/llvm/lib/Transforms/Scalar/CMakeFiles/LLVMScalarOpts.dir/LoopUnrollPass.cpp.o 
/usr/bin/c++ -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/Release64/compiler/llpc/llvm/lib/Transforms/Scalar -I/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/lib/Transforms/Scalar -I/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/Release64/compiler/llpc/llvm/include -I/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG  -fno-exceptions -funwind-tables -fno-rtti -fno-aligned-new -std=c++17 -MD -MT compiler/llpc/llvm/lib/Transforms/Scalar/CMakeFiles/LLVMScalarOpts.dir/LoopUnrollPass.cpp.o -MF compiler/llpc/llvm/lib/Transforms/Scalar/CMakeFiles/LLVMScalarOpts.dir/LoopUnrollPass.cpp.o.d -o compiler/llpc/llvm/lib/Transforms/Scalar/CMakeFiles/LLVMScalarOpts.dir/LoopUnrollPass.cpp.o -c /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp
In file included from /jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/IR/PassManager.h:40,
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llvm-project/llvm/include/llvm/ADT/DenseMap.h:19:10: fatal error: llvm/Support/AlignOf.h: No such file or directory
   19 | #include  llvm/Support/AlignOf.h 
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[1611/3067] Building CXX object compiler/llpc/llvm/lib/Transforms/Scalar/CMakeFiles/LLVMScalarOpts.dir/LoopDeletion.cpp.o
[1612/3067] Building CXX object compiler/llpc/llvm/lib/Transforms/Scalar/CMakeFiles/LLVMScalarOpts.dir/LoopDataPrefetch.cpp.o

File: lgc/patch/PatchBufferOp.cpp [hunk 1]
```diff
@@ -959,6 +960,23 @@ void BufferOpLowering::visitICmpInst(ICmpInst &icmpInst) {
   icmpInst.replaceAllUsesWith(newICmp);
 }
 
+// =====================================================================================================================
+// Visits invariant start intrinsic.
+//
+// @param intrinsic : The intrinsic
+void BufferOpLowering::visitInvariantStart(llvm::IntrinsicInst &intrinsic) {
+  if (intrinsic.getIntrinsicID() != Intrinsic::invariant_start)
+    return;
```
Comment (nhaehnle) [inline lgc/patch/PatchBufferOp.cpp:969]: This shouldn't be needed.

**Review by nhaehnle — COMMENTED** [review]
I would have expected the `m_typeLowering.eraseInstruction(...)` approach to work. The invariant_start should be visited after the `phi`, so it should be erased before the phi. What am I missing?

**Comment by nhaehnle** [comment]
Please simplify the test, it may be best to write a test from scratch that exhibits the problematic pattern.

When I try this branch but use `m_typeLowering.eraseInstruction` instead I do get an error, but it seems to be an assertion unrelated to erasing the instructions?

**Comment by nhaehnle** [comment]
I can't reproduce any issue with the latest branch and the reduced .lgc test. I do see a crash with the pipeline you sent me, though I don't yet understand it.

Why is there an issue with the phis? The `phi` is visited before the `invariant.start`, which means it is added to the `m_instructionsToErase before it. This means the `invariant.start` is erased before the `phi`, so that should be okay.

I do see the crash near trying to erase some `invariant.start`, but it's actually an `invariant.start` that looks unrelated to a phi. I'm still trying to understand this better.

**Comment by ruiminzhao** [comment]
> I can't reproduce any issue with the latest branch and the reduced .lgc test. I do see a crash with the pipeline you sent me, though I don't yet understand it.
> 
> Why is there an issue with the phis? The `phi` is visited before the `invariant.start`, which means it is added to the `m_instructionsToErase before it. This means the `invariant.start`is erased before the`phi`, so that should be okay.
> 
> I do see the crash near trying to erase some `invariant.start`, but it's actually an `invariant.start` that looks unrelated to a phi. I'm still trying to understand this better.


I have synced the latest code, my current reproduction situation is the same with you:
1.  With the reduced .lgc test, the issue can't be reproduced.
2.  Only assert occurs when run amdllpc with the issued pipeline.
====> I have found the cause for the point2: 
 The Intrinsic::invariant_start will be added into m_instructionsToErase  two times:
  1)  First when in removeUsersForInvariantStarts when visit BufferOpLowering::visitBufferDescToPtr .
  2）Second when BufferOpLowering::visitInvariantStart.
 After removing the removeUsersForInvariantStarts  in BufferOpLowering::visitBufferDescToPtr , the issue will be resolved.

You are correct for the order between `invariant.start` and `phi`, please ignore my analysis on this. 

I would like update the PR and *.lgc test later. Thanks.

**Comment by amdvlk-admin** [comment]
### Test summary for commit d4ae52e09d8a731f48aa4f42d743a78043cf17dc
<details><summary>CTS tests (Failed: 0/0)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Timeout or crash</li>

</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
Awesome, looking forward to the update!

**Comment by ruiminzhao** [comment]
Test case and the source code have been updated. Now all the failures have been resolved.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5c5aef39e473e2f297d7494f5d8a62cc841623c1
<details><summary>CTS tests (Failed: 0/142050)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71030 (56.9%)</li>
<li>Failed: 0/71030 (0.0%)</li>
<li>Not Supported: 30604/71030 (43.1%)</li><li>Warnings: 0/71030 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71020 (57.0%)</li>
<li>Failed: 0/71020 (0.0%)</li>
<li>Not Supported: 30556/71020 (43.0%)</li><li>Warnings: 0/71020 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 31440810f50d1a448c08dce5b9734b61deb4afda
<details><summary>CTS tests (Failed: 0/142050)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71030 (56.9%)</li>
<li>Failed: 0/71030 (0.0%)</li>
<li>Not Supported: 30604/71030 (43.1%)</li><li>Warnings: 0/71030 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71020 (57.0%)</li>
<li>Failed: 0/71020 (0.0%)</li>
<li>Not Supported: 30556/71020 (43.0%)</li><li>Warnings: 0/71020 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
Great, thanks! But those Docker fails need another look. If those tests failed initially, then perhaps that's a sign that something still needs to be looked at more closely? Perhaps the invariant setting isn't always applied properly?

**Comment by ruiminzhao** [comment]
> Great, thanks! But those Docker fails need another look. If those tests failed initially, then perhaps that's a sign that something still needs to be looked at more closely? Perhaps the invariant setting isn't always applied properly?

Yes, I'm still investigating on these failures in Docker.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 56a7722e9a1c0bf469c1762be978f8a78d1c3e89
<details><summary>CTS tests (Failed: 0/142050)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71030 (56.9%)</li>
<li>Failed: 0/71030 (0.0%)</li>
<li>Not Supported: 30604/71030 (43.1%)</li><li>Warnings: 0/71030 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71020 (57.0%)</li>
<li>Failed: 0/71020 (0.0%)</li>
<li>Not Supported: 30556/71020 (43.0%)</li><li>Warnings: 0/71020 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request closed without merging**